README.rst
=============================
Plugs Store
=============================

.. image:: https://badge.fury.io/py/plugs-store.svg
    :target: https://badge.fury.io/py/plugs-store

.. image:: https://travis-ci.org/beatrizseabra/plugs-store.svg?branch=master
    :target: https://travis-ci.org/beatrizseabra/plugs-store

.. image:: https://codecov.io/gh/beatrizseabra/plugs-store/branch/master/graph/badge.svg
    :target: https://codecov.io/gh/beatrizseabra/plugs-store

Pluggable Store

Documentation
-------------

The full documentation is at https://plugs-store.readthedocs.io.

Quickstart
----------

Install Plugs Store::

    pip install plugs-store

Add it to your `INSTALLED_APPS`:

.. code-block:: python

    INSTALLED_APPS = (
        ...
        'plugs_store',
        ...
    )

Add Plugs Store's Views to ROUTER:

.. code-block:: python

    from plugs_store import views as store_views

    ROUTER.register(r'items', store_views.ItemViewSet)
    ROUTER.register(r'item-categories', store_views.ItemCategoryViewSet)
    ROUTER.register(r'orders', views.OrderViewSet)
    ROUTER.register(r'order-items', views.OrderItemViewSet)
    ROUTER.register(r'payment-types', views.PaymentTypeViewSet)
    ROUTER.register(r'shipping', views.ShippingViewSet)
    


    urlpatterns = [
        url(r'^', include(ROUTER.urls)),
        ...
    ]

Features
--------

* TODO

Running Tests
-------------

Does the code actually work?

::

    source <YOURVIRTUALENV>/bin/activate
    (myenv) $ pip install tox
    (myenv) $ tox

Credits
-------

Tools used in rendering this package:

*  Cookiecutter_
*  `cookiecutter-djangopackage`_

.. _Cookiecutter: https://github.com/audreyr/cookiecutter
.. _`cookiecutter-djangopackage`: https://github.com/pydanny/cookiecutter-djangopackage


HISTORY.rst
.. :changelog:

History
-------

0.1.0 (2017-07-19)
++++++++++++++++++

* First release on PyPI.


tests/test_models.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-

"""
test_plugs-store
------------

Tests for `plugs-store` models module.
"""

from django.test import TestCase

from plugs_store import models


class TestPlugs_store(TestCase):

    def setUp(self):
        pass

    def test_something(self):
        pass

    def tearDown(self):
        pass


plugs_store/views.py
from rest_framework import permissions
from rest_framework.decorators import permission_classes
from rest_framework.viewsets import ModelViewSet

from plugs_filter.decorators import auto_filters
from plugs_store import permissions as api_permissions
from plugs_store.pagination import ClientRequestedPagination
from plugs_store import serializers, models


@auto_filters
class ItemCategoryViewSet(ModelViewSet):
    """
    Item Category ViewSet
    """
    queryset = models.ItemCategory.objects.all()
    serializer_class = serializers.ItemCategorySerializer
    permission_classes = [api_permissions.StaffOrReadOnly]
    pagination_class = ClientRequestedPagination
    auto_filters_fields = ('is_hidden', 'name')
    ordering_fields = ('id', 'name', 'created', 'updated')


@auto_filters
class ItemViewSet(ModelViewSet):
    """
    Item ViewSet
    """
    queryset = models.Item.objects.all()
    serializer_class = serializers.ItemSerializer
    permission_classes = [api_permissions.StaffOrReadOnly]
    pagination_class = ClientRequestedPagination
    auto_filters_fields = ('name', 'price', 'item_category', 'item_category__is_hidden')
    ordering_fields = ('id', 'name', 'price', 'item_category', 'created', 'updated')


@auto_filters
class OrderViewSet(ModelViewSet):
    """
    Order Viewset
    """
    serializer_class = serializers.OrderSerializer
    permission_classes = [permissions.IsAuthenticated]
    queryset = models.Order.objects.all()
    auto_filters_fields = ('name', 'user', 'state', 'total', 'payment_type', 'payment_id')
    ordering_fields = ('id', 'name', 'user', 'state', 'total', 'payment_id', 'created', 'updated')

    def get_queryset(self):
        """
        View should return a list of orders for the current
        authenticated user
        """
        if not self.request.user.is_staff:
            return self.queryset.filter(user=self.request.user)
        return self.queryset


@auto_filters
class OrderItemViewSet(ModelViewSet):
    """
    Order Items Viewset
    """
    serializer_class = serializers.OrderItemSerializer
    permission_classes = [permissions.IsAuthenticated]
    queryset = models.OrderItem.objects.all()
    auto_filters_fields = ('order', 'item', 'item__item_category', 'price', 'user', 'quantity')
    ordering_fields = ('id', 'order', 'user', 'item', 'price', 'quantity', 'created', 'updated')

    def get_queryset(self):
        """
        View should return list of user's order items
        """
        if not self.request.user.is_staff:
            return self.queryset.filter(user=self.request.user)
        return self.queryset


class PaymentTypeViewSet(ModelViewSet):
    """
    Payment Type ViewSet
    """
    queryset = models.PaymentType.objects.all()
    serializer_class = serializers.PaymentTypeSerializer
    permission_classes = [api_permissions.AdminOrReadOnly]
    pagination_class = ClientRequestedPagination


class ShippingViewSet(ModelViewSet):
    """
    Shipping ViewSet
    """
    queryset = models.Shipping.objects.all()
    serializer_class = serializers.ShippingSerializer
    permission_classes = [permissions.IsAuthenticated]
    pagination_class = ClientRequestedPagination


plugs_store/models.py
# -*- coding: utf-8 -*-
from django.utils import timezone
from django.conf import settings
from django.db import models
from django.db.models.signals import pre_save
from django.dispatch import receiver
from django.utils.translation import ugettext_lazy as _
from django_fsm import FSMField, transition, TransitionNotAllowed

from plugs_core.mixins import Timestampable
from plugs_media.fields import MediaField

#region Store
class ItemCategory(Timestampable):
    """
    Item Category Model
    """
    name = models.CharField(max_length=30, verbose_name=_('name'))
    is_hidden = models.BooleanField(default=False, blank=True)

    def __str__(self):
        """
        Python3 string representation
        """
        return self.name

    class Meta:
        verbose_name = _('item category')
        verbose_name_plural = _('item categories')


class Item(Timestampable):
    """
    Item Model
    """
    name = models.CharField(max_length=50, verbose_name=_('name'))
    description = models.TextField(blank=True, verbose_name=_('description'))
    price = models.DecimalField(max_digits=20, decimal_places=2, verbose_name=_('price'))
    avatar = MediaField(null=True, blank=True, verbose_name=_('avatar'))
    item_category = models.ForeignKey('ItemCategory', null=True, blank=True)

    def __str__(self):
        """
        Python3 string representation
        """
        return self.name

    class Meta:
        verbose_name = _('item')
        verbose_name_plural = _('items')


class PaymentType(Timestampable):
    """
    Type of Payment Model
    """
    id = models.CharField(max_length=10, primary_key=True, unique=True, verbose_name=_('name'))
    name = models.CharField(max_length=25)
    description = models.TextField(blank=True, verbose_name=_('description'))
    avatar = MediaField(null=True, blank=True, verbose_name=_('avatar'))

    def __str__(self):
        """
        Python3 string representation
        """
        return self.name

    class Meta:
        verbose_name = _('payment type')
        verbose_name_plural = _('payment types')


class Shipping(Timestampable):
    """
    Shipping Model
    """
    address = models.TextField(verbose_name=_('address'))
    date = models.DateTimeField(null=True, blank=True, verbose_name=_('date'))
    fee = models.DecimalField(max_digits=6, decimal_places=2, verbose_name=_('fee'))

    def __str__(self):
        """
        Python3 string representation
        """
        return self.address

    class Meta:
        verbose_name = _('shipping')
        verbose_name_plural = _('shipping')


class Order(Timestampable):
    """
    Order Model
    """
    #region states
    CREATED = 'CREATED'
    IN_PAYMENT = 'IN_PAYMENT'
    PAID = 'PAID'
    SHIPPED = 'SHIPPED'
    CANCELED = 'CANCELED'
    INVALID = 'INVALID'

    STATES = [
        (CREATED, _('Created')),
        (IN_PAYMENT, _('In Payment')),
        (PAID, _('Paid')),
        (SHIPPED, _('Shipped')),
        (CANCELED, _('Canceled')),
        (INVALID, _('Invalid')),
    ]
    #endregion

    name = models.CharField(max_length=50, blank=True, verbose_name=_('name'))
    user = models.ForeignKey(settings.AUTH_USER_MODEL)
    total = models.DecimalField(default=0, max_digits=25, decimal_places=2, verbose_name=_('total'))
    payment_type = models.ForeignKey('PaymentType')
    payment_id = models.PositiveIntegerField(null=True, blank=True, verbose_name=_('payment_id'))
    state = FSMField(null=False, blank=True, default=CREATED, choices=STATES, verbose_name=_('state'))
    shipping = models.ForeignKey('Shipping', null=True, blank=True)

    #region transitions
    @transition(field=state, source=[CREATED, IN_PAYMENT], target=INVALID)
    def to_invalid(self):
        pass

    @transition(field=state, source=[CREATED, IN_PAYMENT, PAID], target=CANCELED)
    def to_canceled(self):
        pass

    @transition(field=state, source=PAID, target=SHIPPED)
    def to_shipped(self, data):
        if data is not None:
            self.shipping.date = data
        else:
            self.shipping.date = timezone.now()
        self.shipping.save()

    @transition(field=state, source=[CREATED, IN_PAYMENT], target=PAID)
    def to_paid(self):
        pass

    @transition(field=state, source=CREATED, target=IN_PAYMENT)
    def to_in_payment(self):
        pass
    #endregion

    def get_total(self):
        return self.total

    def checkout_items(self):
        """
        Fetch order item with order null and update it with order id
        """
        order_items = OrderItem.objects.filter(user=self.user, order=None)
        for item in order_items:
            item.order = self
            item.save()

    def compute_total(self):
        self.total = OrderItem.objects.filter(order=self).aggregate(models.Sum('price'))['price__sum']

    def create_payment(self):
        # generate payment instead of payment=None
        payment = None
        if payment is not None:
            self.payment_id = payment.pk
            self.to_in_payment()

    def save(self, *args, **kwargs):
        """
        Overrides save method
        """
        if not self.pk:
            super(Order, self).save(*args, **kwargs)
            self.checkout_items()
            self.compute_total()
            self.create_payment()
            self.save()
        else:
            super(Order, self).save(*args, **kwargs)

    def __str__(self):
        """
        Python3 string representation
        """
        return self.name

    class Meta:
        verbose_name = _('order')
        verbose_name_plural = _('orders')


class OrderItem(Timestampable):
    """
    Order Item Model
    """
    order = models.ForeignKey('Order', null=True, blank=True, verbose_name=_('order'))
    item = models.ForeignKey('Item', verbose_name=_('item'))
    user = models.ForeignKey(settings.AUTH_USER_MODEL, verbose_name=_('user'))
    quantity = models.PositiveIntegerField(default=1, blank=True, verbose_name=_('quantity'))
    price = models.DecimalField(blank=True, max_digits=22, decimal_places=2, verbose_name=_('price'))

    def __str__(self):
        """
        Python3 string representation
        """
        return '{0} ordered'.format(self.item)

    class Meta:
        verbose_name = _('order item')
        verbose_name_plural = _('order items')
#endregion


@receiver(pre_save, sender=OrderItem)
def calculate_order_item_price(instance, sender, **kwargs):
    instance.price = instance.quantity * instance.item.price


plugs_store/serializers.py
from django.utils.translation import ugettext_lazy as _

from rest_framework import serializers
from rest_framework.exceptions import ValidationError

from plugs_store import models


class ItemCategorySerializer(serializers.ModelSerializer):
     """
     Item Category Serializer
     """

     # pylint: disable=R0903
     class Meta:
          """
          Metaclass definition
          """
          model = models.ItemCategory
          fields = ('id', 'name', 'is_hidden', 'created', 'updated')


class ItemSerializer(serializers.ModelSerializer):
     """
     Item Serializer
     """

     # pylint: disable=R0903
     class Meta:
          """
          Metaclass definition
          """
          model = models.Item
          fields = ('id', 'name', 'description', 'price', 'avatar', 'item_category', 'created', 'updated')


class OrderSerializer(serializers.ModelSerializer):
    """
    Order Serializer
    """
    def __init__(self, *args, **kwargs):
        """
        Dealing with extra read_only_fields
        """
        super(OrderSerializer, self).__init__(*args, **kwargs)
        if self.context['request'].method in ['POST']:
            self.fields['state'].read_only = True
        if self.context['request'].user.is_staff:
            self.fields['user'].read_only = False
            self.fields['total'].read_only = False
            self.fields['payment_id'].read_only = False
            self.fields['state'].read_only = False

    def validate(self, data):
        if 'user' not in data:
            data['user'] = self.context['request'].user
        if self.instance is None:
            items = models.OrderItem.objects.filter(user=data['user'], order=None)
            if not items.exists():
                if not self.fields['user'].read_only:
                    raise serializers.ValidationError({'user': [_('You have no items to checkout.')]})
                else:
                    raise serializers.ValidationError({'name': [_('Order cannot be created if user has no items to checkout.')]})
            if 'payment_type' not in data:
                raise serializers.ValidationError({'payment_type': [_('Order cannot be created without choosing a payment type.')]})
        elif not self.context['request'].user.is_staff and 'state' in data and data['state'] != 'CANCELED':
            raise serializers.ValidationError({'state': [_('You can only change order state to canceled.')]})
        return data


    class Meta:
        """
        Metaclass definition
        """
        model = models.Order
        fields = ('id', 'name', 'user', 'total', 'state', 'payment_type', 'payment_id', 'shipping', 'created', 'updated')
        read_only_fields = ('user', 'total', 'payment_id')


class OrderItemSerializer(serializers.ModelSerializer):
    """
    Order Item Serializer
    """
    def __init__(self, *args, **kwargs):
        """
        Dealing with extra read_only_fields
        """
        super(OrderItemSerializer, self).__init__(*args, **kwargs)
        if self.context['request'].user.is_staff:
            self.fields['user'].read_only = False
            self.fields['order'].read_only = False
            self.fields['price'].read_only = False

    def validate(self, data):
        if self.instance is None and 'user' not in data:
            data['user'] = self.context['request'].user
        return data

    class Meta:
        """
        Metaclass definition
        """
        model = models.OrderItem
        fields = ('id', 'order', 'item', 'quantity', 'price', 'user', 'created', 'updated')
        read_only_fields = ('order', 'price', 'user')


class PaymentTypeSerializer(serializers.ModelSerializer):
    """
    Payment Type Serializer
    """

    class Meta:
        """
        Metaclass definition
        """
        model = models.PaymentType
        fields = ('id', 'name', 'description', 'avatar', 'app', 'model', 'created', 'updated')


class ShippingSerializer(serializers.ModelSerializer):
    """
    Shipping Serializer
    """
    def __init__(self, *args, **kwargs):
        """
        Dealing with extra read_only_fields
        """
        super(ShippingSerializer, self).__init__(*args, **kwargs)
        if self.context['request'].user.is_staff:
            self.fields['date'].read_only = False
            self.fields['fee'].read_only = False

    # pylint: disable=R0903
    class Meta:
        """
        Metaclass definition
        """
        model = models.Shipping
        fields = ('id', 'address', 'date', 'fee', 'created', 'updated')
        read_only_fields = ('date', 'fee')


plugs_store/migrations/0001_initial.py
# -*- coding: utf-8 -*-
# Generated by Django 1.11.3 on 2017-07-19 15:18
from __future__ import unicode_literals

from django.conf import settings
from django.db import migrations, models
import django.db.models.deletion
import django_fsm
import plugs_media.fields


class Migration(migrations.Migration):

    initial = True

    dependencies = [
        migrations.swappable_dependency(settings.AUTH_USER_MODEL),
    ]

    operations = [
        migrations.CreateModel(
            name='Item',
            fields=[
                ('id', models.AutoField(auto_created=True, primary_key=True, serialize=False, verbose_name='ID')),
                ('created', models.DateTimeField(auto_now_add=True)),
                ('updated', models.DateTimeField(auto_now=True)),
                ('name', models.CharField(max_length=50, verbose_name='name')),
                ('description', models.TextField(blank=True, verbose_name='description')),
                ('price', models.DecimalField(decimal_places=2, max_digits=20, verbose_name='price')),
                ('avatar', plugs_media.fields.MediaField(blank=True, max_length=100, null=True, verbose_name='avatar')),
            ],
            options={
                'verbose_name': 'item',
                'verbose_name_plural': 'items',
            },
        ),
        migrations.CreateModel(
            name='ItemCategory',
            fields=[
                ('id', models.AutoField(auto_created=True, primary_key=True, serialize=False, verbose_name='ID')),
                ('created', models.DateTimeField(auto_now_add=True)),
                ('updated', models.DateTimeField(auto_now=True)),
                ('name', models.CharField(max_length=30, verbose_name='name')),
                ('is_hidden', models.BooleanField(default=False)),
            ],
            options={
                'verbose_name': 'item category',
                'verbose_name_plural': 'item categories',
            },
        ),
        migrations.CreateModel(
            name='Order',
            fields=[
                ('id', models.AutoField(auto_created=True, primary_key=True, serialize=False, verbose_name='ID')),
                ('created', models.DateTimeField(auto_now_add=True)),
                ('updated', models.DateTimeField(auto_now=True)),
                ('name', models.CharField(blank=True, max_length=50, verbose_name='name')),
                ('total', models.DecimalField(decimal_places=2, default=0, max_digits=25, verbose_name='total')),
                ('payment_id', models.PositiveIntegerField(blank=True, null=True, verbose_name='payment_id')),
                ('state', django_fsm.FSMField(blank=True, choices=[('CREATED', 'Created'), ('IN_PAYMENT', 'In Payment'), ('PAID', 'Paid'), ('SHIPPED', 'Shipped'), ('CANCELED', 'Canceled'), ('INVALID', 'Invalid')], default='CREATED', max_length=50, verbose_name='state')),
            ],
            options={
                'verbose_name': 'order',
                'verbose_name_plural': 'orders',
            },
        ),
        migrations.CreateModel(
            name='OrderItem',
            fields=[
                ('id', models.AutoField(auto_created=True, primary_key=True, serialize=False, verbose_name='ID')),
                ('created', models.DateTimeField(auto_now_add=True)),
                ('updated', models.DateTimeField(auto_now=True)),
                ('quantity', models.PositiveIntegerField(blank=True, default=1, verbose_name='quantity')),
                ('price', models.DecimalField(blank=True, decimal_places=2, max_digits=22, verbose_name='price')),
                ('item', models.ForeignKey(on_delete=django.db.models.deletion.CASCADE, to='plugs_store.Item', verbose_name='item')),
                ('order', models.ForeignKey(blank=True, null=True, on_delete=django.db.models.deletion.CASCADE, to='plugs_store.Order', verbose_name='order')),
                ('user', models.ForeignKey(on_delete=django.db.models.deletion.CASCADE, to=settings.AUTH_USER_MODEL, verbose_name='user')),
            ],
            options={
                'verbose_name': 'order item',
                'verbose_name_plural': 'order items',
            },
        ),
        migrations.CreateModel(
            name='PaymentType',
            fields=[
                ('created', models.DateTimeField(auto_now_add=True)),
                ('updated', models.DateTimeField(auto_now=True)),
                ('id', models.CharField(max_length=10, primary_key=True, serialize=False, unique=True, verbose_name='name')),
                ('name', models.CharField(max_length=25)),
                ('description', models.TextField(blank=True, verbose_name='description')),
                ('avatar', plugs_media.fields.MediaField(blank=True, max_length=100, null=True, verbose_name='avatar')),
            ],
            options={
                'verbose_name': 'payment type',
                'verbose_name_plural': 'payment types',
            },
        ),
        migrations.CreateModel(
            name='Shipping',
            fields=[
                ('id', models.AutoField(auto_created=True, primary_key=True, serialize=False, verbose_name='ID')),
                ('created', models.DateTimeField(auto_now_add=True)),
                ('updated', models.DateTimeField(auto_now=True)),
                ('address', models.TextField(verbose_name='address')),
                ('date', models.DateTimeField(blank=True, null=True, verbose_name='date')),
                ('fee', models.DecimalField(decimal_places=2, max_digits=6, verbose_name='fee')),
            ],
            options={
                'verbose_name': 'shipping',
                'verbose_name_plural': 'shipping',
            },
        ),
        migrations.AddField(
            model_name='order',
            name='payment_type',
            field=models.ForeignKey(on_delete=django.db.models.deletion.CASCADE, to='plugs_store.PaymentType'),
        ),
        migrations.AddField(
            model_name='order',
            name='shipping',
            field=models.ForeignKey(blank=True, null=True, on_delete=django.db.models.deletion.CASCADE, to='plugs_store.Shipping'),
        ),
        migrations.AddField(
            model_name='order',
            name='user',
            field=models.ForeignKey(on_delete=django.db.models.deletion.CASCADE, to=settings.AUTH_USER_MODEL),
        ),
        migrations.AddField(
            model_name='item',
            name='item_category',
            field=models.ForeignKey(blank=True, null=True, on_delete=django.db.models.deletion.CASCADE, to='plugs_store.ItemCategory'),
        ),
    ]
